File: Rust/2020/06.rs
#![feature(test)]

use rustc_hash::FxHashSet;

type Input = Vec<Vec<Vec<u8>>>;

fn setup(input: &str) -> Input {
    input
        .split("\n\n")
        .map(|group| group.lines().map(|line| line.bytes().collect()).collect())
        .collect()
}

fn part1(input: &Input) -> usize {
    input
        .iter()
        .map(|group| group.iter().flatten().collect::<FxHashSet<_>>().len())
        .sum()
}

fn part2(input: &Input) -> usize {
    input
        .iter()
        .map(|group| {
            let l = group.len();
            group
                .iter()
                .fold([0; 256], |mut acc, xs| {
                    for &x in xs {
                        acc[x as usize] += 1;
                    }
                    acc
                })
                .into_iter()
                .filter(|&x| x == l)
                .count()
        })
        .sum()
}

aoc::main!(2020, 6, ex: 1);

File: Python/2020/06.py
from lib import *

input = read_input(2020, 6)


out = 0
for group in input.split("\n\n"):
    out += len(set(group.replace("\n", "")))

print(out)


out = 0
for group in input.split("\n\n"):
    x = set(group.replace("\n", ""))
    for person in group.split():
        x &= set(person)
    out += len(x)

print(out)
